sentry-android-core/src/main/java/io/sentry/android/core/AndroidUserCache.java

```diff
@@ -0,0 +1,90 @@
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+// TODO: to think, this could be a SentryOptionsCache that could be expanded, does it make sense?
+
+/** the Android User Cache class. It caches the user data in the Android SharedPreferences */
+public final class AndroidUserCache implements IUserCache {
+
+  // TODO: only these 3 fields make sense to me, otherwise caching a raw json would be better, but I
+  // don't want to do that.
+  private static final String EMAIL = "email";
+  private static final String ID = "id";
+  private static final String USERNAME = "user_name";
+
+  private final @NotNull SharedPreferences sharedPreferences;
+  private final @NotNull SentryAndroidOptions options;
+  private final @Nullable String sentryDeviceId;
+
+  public AndroidUserCache(
+      final @NotNull SentryAndroidOptions options,
+      final @NotNull SharedPreferences sharedPreferences,
+      final @Nullable String sentryDeviceId) {
+    this.options = Objects.requireNonNull(options, "SentryAndroidOptions is required.");
+    this.sharedPreferences =
+        Objects.requireNonNull(sharedPreferences, "SharedPreferences is required.");
+    this.sentryDeviceId = sentryDeviceId;
+  }
+
+  //  TODO: I believe this makes sense only for global hub SDKs, but if not,
+  //   how do we cache multiple users (because we have multiple scopes)?
+
+  @Override
+  public void setUser(final @Nullable User user) {
+    if (options.isCacheUserForSessions()) {
+      final SharedPreferences.Editor edit = sharedPreferences.edit();
+      if (user != null) {
+        edit.putString(ID, user.getId())
+            .putString(EMAIL, user.getEmail())
+            .putString(USERNAME, user.getUsername())
+            .apply(); // it does in-memory cache and flush to the disk async.
+      } else {
+        cleanUserFields(edit);
+      }
+    }
+  }
+
+  @Override
+  public @Nullable User getUser() {
+    // sharedPreferences has in-memory cache and flush to the disk async., so it's not expensive
+    if (options.isCacheUserForSessions()) {
+      final String id = sharedPreferences.getString(ID, null);
+      final String email = sharedPreferences.getString(EMAIL, null);
+      final String userName = sharedPreferences.getString(USERNAME, null);
+
+      boolean hasCachedUser = true;
+      if (id == null && email == null && userName == null) {
+        hasCachedUser = false;
+      }
+
+      final User user = new User();
+
+      if (hasCachedUser) {
+        // returns previous set user, even if it has no id
+        user.setId(id);
+        user.setEmail(email);
+        user.setUsername(userName);
+      } else {
+        user.setId(sentryDeviceId);
+      }
+      return user;
+    } else {
+      final SharedPreferences.Editor edit = sharedPreferences.edit();
+      // is it a good idea to do it here? this guarantees that it cleans up on restart if the flag
+      // has been swapped
+      cleanUserFields(edit);
```

Comment:
to discuss

Reply:
I think it makes sense, because otherwise you could receive an old user. Imagine the following scenario:
1. `setUser(A)`
2. `cacheUserForSessions = false`
3. `setUser(B)`, which doesn't do anything.
4. `cacheUserForSessions = true`
5. `getUser()` returns `A`

Reply:
yep, that was exactly the use-case

Reply:
I just came up with another scenario:
1. `setUser(A)`
2. cacheUserForSessions = false
3. cacheUserForSessions = true
4. getUser() returns A

In this scenario, even with calling `cleanUserFields` in `setUser` and `getUser`, you get the old user A. `SentryOptions.setCacheUserForSessions` could clear the `UserCache` if set to `false` to solve this problem.

Reply:
yeah, I'd say on SDK. init, if the flag is `false`, we should clean this up.
SentryOptions fields are only mutable on SDK init., but not on runtime.
good catch